File: src/main/java/org/taktik/icure/services/external/rest/v1/dto/filter/contact/ContactByHcPartyTagCodeDateFilter.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.services.external.rest.v1.dto.filter.contact;

import com.google.common.base.Objects;
import org.taktik.icure.entities.Contact;
import org.taktik.icure.services.external.rest.handlers.JsonPolymorphismRoot;
import org.taktik.icure.services.external.rest.v1.dto.filter.Filter;

import java.util.List;

@JsonPolymorphismRoot(Filter.class)
public class ContactByHcPartyTagCodeDateFilter extends Filter<Contact> implements org.taktik.icure.dto.filter.contact.ContactByHcPartyTagCodeDateFilter {
	String healthcarePartyId;
	String tagType;
	String tagCode;
	String codeType;
	String codeCode;
	Long startOfContactOpeningDate;
	Long endOfContactOpeningDate;

	public ContactByHcPartyTagCodeDateFilter() {
	}

	public ContactByHcPartyTagCodeDateFilter(String healthcarePartyId, List<String> patientSecretForeignKeys, String tagType, String tagCode, String codeType, String codeCode, Long startOfContactOpeningDate, Long endOfContactOpeningDate) {
		this.healthcarePartyId = healthcarePartyId;
		this.tagType = tagType;
		this.tagCode = tagCode;
		this.codeType = codeType;
		this.codeCode = codeCode;
		this.startOfContactOpeningDate = startOfContactOpeningDate;
		this.endOfContactOpeningDate = endOfContactOpeningDate;
	}

	@Override
	public String getHealthcarePartyId() {
		return healthcarePartyId;
	}

	public void setHealthcarePartyId(String healthcarePartyId) {
		this.healthcarePartyId = healthcarePartyId;
	}

    @Override
	public String getTagType() {
		return tagType;
	}

	public void setTagType(String tagType) {
		this.tagType = tagType;
	}

	@Override
	public String getTagCode() {
		return tagCode;
	}

	public void setTagCode(String tagCode) {
		this.tagCode = tagCode;
	}

	@Override
	public String getCodeType() {
		return codeType;
	}

	public void setCodeType(String codeType) {
		this.codeType = codeType;
	}

	@Override
	public String getCodeCode() {
		return codeCode;
	}

	public void setCodeCode(String codeCode) {
		this.codeCode = codeCode;
	}

	public Long getStartOfContactOpeningDate() {
		return startOfContactOpeningDate;
	}

	public void setStartOfContactOpeningDate(Long startOfContactOpeningDate) {
		this.startOfContactOpeningDate = startOfContactOpeningDate;
	}

	public Long getEndOfContactOpeningDate() {
		return endOfContactOpeningDate;
	}

	public void setEndOfContactOpeningDate(Long endOfContactOpeningDate) {
		this.endOfContactOpeningDate = endOfContactOpeningDate;
	}

	@Override
	public int hashCode() {
		return Objects.hashCode(healthcarePartyId, tagType, tagCode, codeType, codeCode, startOfContactOpeningDate, endOfContactOpeningDate);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null || getClass() != obj.getClass()) {
			return false;
		}
		final ContactByHcPartyTagCodeDateFilter other = (ContactByHcPartyTagCodeDateFilter) obj;
		return Objects.equal(this.healthcarePartyId, other.healthcarePartyId) && Objects.equal(this.tagType, other.tagType) && Objects.equal(this.tagCode, other.tagCode) && Objects.equal(this.codeType, other.codeType) && Objects.equal(this.codeCode, other.codeCode) && Objects.equal(this.startOfContactOpeningDate, other.startOfContactOpeningDate) && Objects.equal(this.endOfContactOpeningDate, other.endOfContactOpeningDate);
	}

	@Override
	public boolean matches(Contact item) {
		return (healthcarePartyId == null || item.getDelegations().keySet().contains(healthcarePartyId))
				&& (tagType == null || item.getServices().stream().filter(s ->
				(s.getTags().stream().filter(t -> tagType.equals(t.getType()) && (tagCode == null || tagCode.equals(t.getCode()))).findAny().isPresent())
						&& (codeType == null || (s.getCodes().stream().filter(c -> codeType.equals(c.getType()) && (codeCode == null || codeCode.equals(c.getCode()))).findAny().isPresent()))
						&& (startOfContactOpeningDate == null || (s.getValueDate() != null && s.getValueDate() > startOfContactOpeningDate) || (s.getOpeningDate() != null && s.getOpeningDate() > startOfContactOpeningDate))
						&& (endOfContactOpeningDate == null || (s.getValueDate() != null && s.getValueDate() < endOfContactOpeningDate) || (s.getOpeningDate() != null && s.getOpeningDate() < endOfContactOpeningDate))
		).findAny().isPresent());
	}
}

File: src/main/kotlin/org/taktik/icure/domain/filter/impl/contact/ContactByHcPartyTagCodeDateFilter.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.domain.filter.impl.contact

import com.github.pozo.KotlinBuilder
import com.google.common.base.Objects
import org.taktik.icure.domain.filter.AbstractFilter
import org.taktik.icure.entities.Contact

@KotlinBuilder
data class ContactByHcPartyTagCodeDateFilter(
	override val desc: String? = null,
	override val healthcarePartyId: String? = null,
	override val tagType: String? = null,
	override val tagCode: String? = null,
	override val codeType: String? = null,
	override val codeCode: String? = null,
	override val startOfContactOpeningDate: Long? = null,
	override val endOfContactOpeningDate: Long? = null
) : AbstractFilter<Contact>, org.taktik.icure.domain.filter.contact.ContactByHcPartyTagCodeDateFilter {
	override fun hashCode(): Int {
		return Objects.hashCode(healthcarePartyId, tagType, tagCode, codeType, codeCode, startOfContactOpeningDate, endOfContactOpeningDate)
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) {
			return true
		}
		if (other == null || javaClass != other.javaClass) {
			return false
		}
		val filter = other as ContactByHcPartyTagCodeDateFilter
		return Objects.equal(healthcarePartyId, filter.healthcarePartyId) && Objects.equal(tagType, filter.tagType) && Objects.equal(tagCode, filter.tagCode) && Objects.equal(codeType, filter.codeType) && Objects.equal(codeCode, filter.codeCode) && Objects.equal(startOfContactOpeningDate, filter.startOfContactOpeningDate) && Objects.equal(endOfContactOpeningDate, filter.endOfContactOpeningDate)
	}

	override fun matches(item: Contact): Boolean {
		return (
			(healthcarePartyId == null || item.delegations.keys.contains(healthcarePartyId!!)) &&
				(
					tagType == null || item.services.any { svc ->
						(
							svc.tags.any { t -> tagType == t.type && (tagCode == null || tagCode == t.code) } &&
								(codeType == null || svc.codes.any { cs -> codeType == cs.type && (codeCode == null || codeCode == cs.code) }) &&
								(startOfContactOpeningDate == null || svc.valueDate != null && svc.valueDate > startOfContactOpeningDate!! || svc.openingDate != null && svc.openingDate > startOfContactOpeningDate!!) &&
								(endOfContactOpeningDate == null || svc.valueDate != null && svc.valueDate < endOfContactOpeningDate!! || svc.openingDate != null && svc.openingDate < endOfContactOpeningDate!!)
							)
					}
					)
			)
	}
}
